In [2]:
import pandas as pd
from datasets import load_dataset
import matplotlib.pyplot as plt
#importing libraries

dataset = load_dataset('lukebarousse/data_jobs')
df = dataset['train'].to_pandas()
# loading data

df['job_posted_date'] = pd.to_datetime(df['job_posted_date'])
#data cleanup

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 785741 entries, 0 to 785740
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   job_title_short        785741 non-null  object        
 1   job_title              785740 non-null  object        
 2   job_location           784696 non-null  object        
 3   job_via                785733 non-null  object        
 4   job_schedule_type      773074 non-null  object        
 5   job_work_from_home     785741 non-null  bool          
 6   search_location        785741 non-null  object        
 7   job_posted_date        785741 non-null  datetime64[ns]
 8   job_no_degree_mention  785741 non-null  bool          
 9   job_health_insurance   785741 non-null  bool          
 10  job_country            785692 non-null  object        
 11  salary_rate            33067 non-null   object        
 12  salary_year_avg        22003 non-null   floa

For this, we will be cleaning up the dataframe. 
- We will replace the NA values with a median value
- We will also look to remove any duplicate job entries (these may be the same jobs but listed on different job websites)

Note: everything we do here will not necessarily be applied to the final dataframe we are working on

In [6]:
df.loc[0:10, 'salary_year_avg':'salary_hour_avg']

salary_year_avg  salary_hour_avg
0               NaN              NaN
1               NaN              NaN
2               NaN              NaN
3               NaN              NaN
4               NaN              NaN
5               NaN              NaN
6               NaN              NaN
7               NaN              NaN
8               NaN              NaN
9               NaN              NaN
10              NaN              NaN

### Replacing NaN with a given value

To replace 'NaN' rows in the filtered df above with median values, we first have to calculate the  median values for each column and set these equal to a variable:

In [9]:
median_salary_year = df['salary_year_avg'].median()

In [10]:
median_salary_hour = df['salary_hour_avg'].median()

To replace NaN without replacing the current values that are in there, we can use the:

- df.fillna(value) method

where 'value' is the value you use to fill in the holes. Here, we will add the respective variables 

In [14]:
df_filled = df
df_filled['salary_year_avg'].fillna(median_salary_year)

0         115000.0
1         115000.0
2         115000.0
3         115000.0
4         115000.0
            ...   
785736    115000.0
785737    115000.0
785738    115000.0
785739    115000.0
785740    115000.0
Name: salary_year_avg, Length: 785741, dtype: float64

In [15]:
df_filled ['salary_year_avg']


0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
          ..
785736   NaN
785737   NaN
785738   NaN
785739   NaN
785740   NaN
Name: salary_year_avg, Length: 785741, dtype: float64

However, we see that there are still NaN values. This is because we need to update this column by making it equal to itself during the df.fillna() method. Therefore, for both 'salary_year_avg' and 'salary_hour_avg':

In [18]:
df_filled['salary_year_avg'] = df_filled['salary_year_avg'].fillna(median_salary_year)
df_filled['salary_hour_avg'] = df_filled['salary_hour_avg'].fillna(median_salary_hour)

In [24]:
df_filled.loc[:10, 'salary_year_avg': 'salary_hour_avg']

salary_year_avg  salary_hour_avg
0          115000.0            45.98
1          115000.0            45.98
2          115000.0            45.98
3          115000.0            45.98
4          115000.0            45.98
5          115000.0            45.98
6          115000.0            45.98
7          115000.0            45.98
8          115000.0            45.98
9          115000.0            45.98
10         115000.0            45.98

### Removing duplicate job entries

For this, we can use the method:

- df.drop_duplicates(subset = )

whereby 'subset' is used to specify a column label or sequence of column labels. This returns a dataframe with duplicate rows removed only IF the values are the same in all of the specified columns in the subset parameter.

In [32]:
df_unique = df_filled
df_unique = df_unique.drop_duplicates()

In [33]:
print('Number of rows', len(df_filled))
print('Number of non-duplicate rows', len(df_unique))
print('Number of dropped rows:', len(df_filled)-len(df_unique))

Number of rows 785741
Number of non-duplicate rows 785640
Number of dropped rows: 101


 Here, we used the drop_duplicates() method for this and by default, it drops rows that are identical, **meaning the dropped rows contain values in all columns that are the same.**

 By specifying the subset, we could go further and select certain columns to filter for any duplicates. Here, we are going to the filter by the:
 - job_title column
 - company_name

In [34]:
df_unique = df_unique.drop_duplicates(subset = ['job_title', 'company_name'])

print('Number of rows', len(df_filled))
print('Number of non-duplicate rows', len(df_unique))
print('Number of dropped rows:', len(df_filled)-len(df_unique))

Number of rows 785741
Number of non-duplicate rows 508042
Number of dropped rows: 277699


Here, we dropped ~277,699 jobs from different companies. Note: in this example, it only dropped rows where values were the same in both of the columns 'job_title' and 'company_name'.